statsmodels/tsa/vector_ar/vecm.py

```diff
@@ -0,0 +1,2247 @@
+        (dimensions assuming no deterministic terms are given)
+        Lagged differenced endog, used as regressor for the short term
+        equation.
+
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+
+    """
+    # p. 286:
+    p = diff_lags+1
+    y = endog
+    K = y.shape[0]
+    y_1_T = y[:, p:]
+    T = y_1_T.shape[1]
+    delta_y = np.diff(y)
+    delta_y_1_T = delta_y[:, p-1:]
+
+    y_lag1 = y[:, p-1:-1]
+    if "co" in deterministic and "ci" in deterministic:
+        raise ValueError("Both 'co' and 'ci' as deterministic terms given. " +
+                         "Please choose one of the two.")
+    y_lag1_stack = [y_lag1]
+    if "ci" in deterministic:  # pp. 257, 299, 306, 307
+        y_lag1_stack.append(np.ones(T))
+    if "li" in deterministic:  # p. 299
+        y_lag1_stack.append(_linear_trend(T, p, coint=True))
+    if exog_coint is not None:
+        y_lag1_stack.append(exog_coint[-T-1:-1].T)
+    y_lag1 = np.row_stack(y_lag1_stack)
+
+    # p. 286:
+    delta_x = np.zeros((diff_lags*K, T))
+    if diff_lags > 0:
+        for j in range(delta_x.shape[1]):
+            delta_x[:, j] = (delta_y[:, j+p-2:None if j-1 < 0 else j-1:-1]
+                             .T.reshape(K*(p-1)))
+    delta_x_stack = [delta_x]
+    # p. 299, p. 303:
+    if "co" in deterministic:
+        delta_x_stack.append(np.ones(T))
+    if seasons > 0:
+        delta_x_stack.append(seasonal_dummies(seasons, delta_x.shape[1],
+                                              first_period=first_season,
+                                              centered=True).T)
+    if "lo" in deterministic:
+        delta_x_stack.append(_linear_trend(T, p))
+    if exog is not None:
+        delta_x_stack.append(exog[-T:].T)
+    delta_x = np.row_stack(delta_x_stack)
+
+    return y_1_T, delta_y_1_T, y_lag1, delta_x
+
+
+def _r_matrices(delta_y_1_T, y_lag1, delta_x):
+    """Returns two ndarrays needed for parameter estimation as well as the
+    calculation of standard errors.
+
+    Parameters
+    ----------
+    delta_y_1_T : ndarray (neqs x nobs)
+        The first differences of endog.
+        `.. math:: (y_1, \\ldots, y_T) - (y_0, \\ldots, y_{T-1})
+    y_lag1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        Endog of the previous period (lag 1).
+        `.. math:: (y_0, \\ldots, y_{T-1})
+    delta_x : ndarray (k_ar_diff*neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        Lagged differenced endog, used as regressor for the short term
+        equation.
+
+    Returns
+    -------
+    result : tuple
+        A tuple of two ndarrays. (See p. 292 in [1]_ for the definition of
+        R_0 and R_1.)
+
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
+
+    # todo: rewrite m such that a big (TxT) matrix is avoided
```

Comment:
Can the nobs x nobs problem be avoided by separating out the left-multiplication by `delta_x.T`?:

```
not_m = inv(delta_x.dot(delta_x.T)).dot(delta_x) # shape --> (k_ar_diff*neqs, nobs)
pre_r0 = delta_y_1_T.dot(delta_x.T)              # shape --> (neqs, k_ar_diff*neqs)
pre_r1 = y_lag1.dot(delta_x.T)                   # shape --> (neqs, k_ar_diff*neqs)
r0 = delta_y_1_T - pre_r0.dot(not_m)
r1 = y_lag1 - pre_r1.dot(not_m)
```